File: app/src/main/java/com/amitbharat/keyboard/utils/LocaleHelper.java
package com.amitbharat.keyboard.utils;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.os.Build;
import android.os.LocaleList;
import androidx.preference.PreferenceManager;
import java.util.Locale;

public class LocaleHelper {

    private static final String PREF_LOCALE = "pref_app_locale";

    public static Context wrapContext(Context context) {
        String lang = getLanguage(context);
        return updateResources(context, lang);
    }

    public static String getLanguage(Context context) {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        return prefs.getString(PREF_LOCALE, Locale.getDefault().getLanguage());
    }

    public static boolean isHindi(Context context) {
        return "hi".equalsIgnoreCase(getLanguage(context));
    }

    public static void setLocale(Context context, String language) {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        prefs.edit().putString(PREF_LOCALE, language).commit();

        updateResources(context, language);

        androidx.appcompat.app.AppCompatDelegate.setApplicationLocales(
                androidx.core.os.LocaleListCompat.forLanguageTags(language)
        );

        if (context instanceof android.app.Activity) {
            ((android.app.Activity) context).recreate();
        }
    }

    private static Context updateResources(Context context, String language) {
        Locale locale = new Locale(language);
        Locale.setDefault(locale);

        Resources resources = context.getResources();
        Configuration configuration = new Configuration(resources.getConfiguration());

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            configuration.setLocales(new LocaleList(locale));
            return context.createConfigurationContext(configuration);
        } else {
            configuration.locale = locale;
            resources.updateConfiguration(configuration, resources.getDisplayMetrics());
            return context;
        }
    }
}

File: app/src/main/java/com/amitbharat/hindikeyboard/utils/LocaleHelper.kt
package com.amitbharat.hindikeyboard.utils

import android.app.Activity
import android.content.Context
import android.content.SharedPreferences
import android.content.res.Configuration
import android.content.res.Resources
import android.os.Build
import android.os.LocaleList
import java.util.Locale

object LocaleHelper {

    private const val PREFS_NAME = "app_language_prefs"
    private const val KEY_LANGUAGE = "selected_language"

    fun isHindi(context: Context): Boolean {
        return getLanguage(context).startsWith("hi")
    }

    fun getLanguage(context: Context): String {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        return prefs.getString(KEY_LANGUAGE, "en") ?: "en"
    }

    fun wrapContext(context: Context): Context {
        val lang = getLanguage(context)
        val locale = Locale(lang)
        Locale.setDefault(locale)

        val res = context.resources
        val config = Configuration(res.configuration)

        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            config.setLocale(locale)
            val localeList = LocaleList(locale)
            LocaleList.setDefault(localeList)
            config.setLocales(localeList)
            context.createConfigurationContext(config)
        } else {
            @Suppress("DEPRECATION")
            config.locale = locale
            @Suppress("DEPRECATION")
            res.updateConfiguration(config, res.displayMetrics)
            context
        }
    }

    fun applyAppLanguage(context: Context) {
        // wrapContext(Context) in attachBaseContext properly handles per-app locale
        // on Android N+ (API 24+) without triggering infinite activity recreation loops.
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
            val lang = getLanguage(context)
            val locale = Locale(lang)
            Locale.setDefault(locale)
            val res = context.resources
            val config = Configuration(res.configuration)
            @Suppress("DEPRECATION")
            config.locale = locale
            @Suppress("DEPRECATION")
            res.updateConfiguration(config, res.displayMetrics)
        }
    }

    fun setLocale(activity: Activity, languageTag: String) {
        val prefs = activity.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        prefs.edit().putString(KEY_LANGUAGE, languageTag).apply()

        val locale = Locale(languageTag)
        Locale.setDefault(locale)

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
            val res = activity.resources
            val config = Configuration(res.configuration)
            @Suppress("DEPRECATION")
            config.locale = locale
            @Suppress("DEPRECATION")
            res.updateConfiguration(config, res.displayMetrics)
            @Suppress("DEPRECATION")
            activity.applicationContext.resources.updateConfiguration(config, activity.applicationContext.resources.displayMetrics)
        }

        try {
            androidx.appcompat.app.AppCompatDelegate.setApplicationLocales(
                androidx.core.os.LocaleListCompat.forLanguageTags(languageTag)
            )
        } catch (_: Exception) {
        }

        activity.recreate()
    }
}
